plugins/zenoh-plugin-trait/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    fn is_compatible_with(others: &[Compatibility]) -> Result<Compatibility, Incompatibility> {
+        let own_compatibility = Self::compatibility();
+        if others.iter().any(|c| c == &own_compatibility) {
+            let conflicting_with = own_compatibility.clone();
+            Err(Incompatibility {
+                own_compatibility,
+                conflicting_with,
+                details: None,
+            })
+        } else {
+            Ok(own_compatibility)
+        }
+    }
+
+    /// Returns the arguments that are required for the plugin's construction
+    fn get_expected_args() -> Vec<Arg<'static, 'static>>;
+
+    /// Constructs an instance of the plugin, which will be launched with [PluginLaunch::start]
+    fn init(args: &ArgMatches) -> Result<Self, Box<dyn Error>>;
+
+    fn box_init(args: &ArgMatches) -> Result<Box<dyn PluginLaunch>, Box<dyn Error>> {
+        match Self::init(args) {
+            Ok(v) => Ok(Box::new(v)),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+/// Allows a [Plugin] instance to be started.
+pub trait PluginLaunch {
+    fn start(self, runtime: Runtime) -> Box<dyn PluginStopper>;
+}
+
+/// Allows a [Plugin] instance to be stopped.
+/// Typically, you can achieve this using a one-shot channel or an [AtomicBool](std::sync::atomic::AtomicBool).
+/// If you don't want a stopping mechanism, you can use `()` as your [PluginStopper].
+pub trait PluginStopper {
+    fn stop(self);
+}
+
+impl PluginStopper for () {
+    fn stop(self) {}
+}
+
+pub mod dynamic_loading {
+    use super::*;
+    pub use no_mangle::*;
+
+    type InitFn = fn(&ArgMatches) -> Result<Box<dyn PluginLaunch>, Box<dyn Error>>;
+    pub type PluginVTableVersion = u16;
+
+    /// This number should change any time the internal structure of [PluginVTable] changes
+    pub const PLUGIN_VTABLE_VERSION: PluginVTableVersion = 0;
+
+    #[repr(C)]
+    struct PluginVTableInner {
+        init: InitFn,
+        is_compatible_with: fn(&[Compatibility]) -> Result<Compatibility, Incompatibility>,
+        get_expected_args: fn() -> Vec<Arg<'static, 'static>>,
+    }
+
+    /// Automagical padding such that [PluginVTable::init]'s result is the size of a cache line
+    #[repr(C)]
+    struct PluginVTablePadding {
+        __padding: [u8; PADDING_LENGTH],
+    }
+    const PADDING_LENGTH: usize =
+        64 - std::mem::size_of::<Result<PluginVTableInner, PluginVTableVersion>>();
+    impl PluginVTablePadding {
+        fn new() -> Self {
+            PluginVTablePadding {
+                __padding: [0; PADDING_LENGTH],
+            }
+        }
+    }
+
+    /// For use with dynamically loaded plugins. Its size will not change accross versions, but its internal structure might.
+    ///
+    /// To ensure compatibility, its size and alignment must allow `size_of::<Result<PluginVTable, PluginVTableVersion>>() == 64` (one cache line).
```

Comment:
size_of::<Result<**PluginVTable**, PluginVTableVersion>> or size_of::<Result<**PluginVTableInner**, PluginVTableVersion>> ?

Reply:
**PluginVTable**, the goal is for the size of the outer type to be fixed even if we change the size of the inner one (by adding more functions to the vtable for example).

The constant should let us know that the pointers may not point to the same things, but I'm afraid of what would happen if the load_plugin function returns something bigger (or smaller) than what is expected by the caller.